Controllers/Frontend/StripePayment.php

```diff
@@ -333,34 +482,67 @@ protected function processChargeSucceededEvent(Stripe\Event $event)
+     * Tries to find the order the event belongs to and, if found, update its payment status
+     * to 'review necessary'.
      *
      * @param Stripe\Event $event
      */
-    protected function processSourceChargeableEvent(Stripe\Event $event)
+    protected function processSourceFailedEvent(Stripe\Event $event)
     {
-        // Check whether the webhook event is allowed to create an order
-        $source = $event->data->object;
-        $stripeSession = Util::getStripeSession();
-        if ($source->id !== $stripeSession->processingSourceId) {
+        $order = $this->findOrderForWebhookEvent($event);
+        if (!$order) {
             return;
         }
+        $paymentStatus = $this->get('models')->find(Status::class, Status::PAYMENT_STATE_REVIEW_NECESSARY);
+        $order->setPaymentStatus($paymentStatus);
+        $this->get('models')->flush($order);
+    }
 
+    /**
+     * First wait for five seconds to prevent timing issues caused by webhooks arriving
+     * earlier than e.g. a redirect during the payment process. That is, if completing the payment
+     * process involves e.g. a redirect to the payment provider, the 'source.chargeable' event might
+     * arrive at the shop earlier than the redirect returns. By pausing the webhook handler, we give the redirect a
+     * head start to complete the order creation. After waiting, the database is checked for an
+     * order that used the event's source. If an order is found and its transactionId starts with 'src_pending',
+     * the source and order are used to create a charge and the order gets updated with the charge.
+     * If no such order is found, check the Shopware session for the 'stripePayment->processingSourceId' field and
+     * make sure the ID matches the source contained in the event because we need the users session to
+     * create the charge and persist the order.
+     *
+     * @param Stripe\Event $event
+     */
+    protected function processSourceChargeableEvent(Stripe\Event $event)
+    {
         // Wait for five seconds
         sleep(5);
 
+        // Retrieve the source from the stripe event
+        $source = $event->data->object;
+
         // Make sure the source has not already been used to create an order, e.g. by completing
         // a redirect
         $order = $this->findOrderForWebhookEvent($event);
         if ($order) {
+            if (mb_substr($order->getTransactionId(), 0, 11) !== 'src_pending') {
+                return;
+            }
+            $charge = $this->createChargeForOrder($source, $order);
+            $order = $this->updateOrderWithCharge($charge);
+            $this->get('pluginlogger')->info(
+                'StripePayment: Updated order after receiving "source.chargeable" webhook event',
+                [
+                    'orderId' => $order->getId(),
+                    'eventId' => $event->id,
+                ]
+            );
```

Comment:
We should log first before doing any updates:
```suggestion
            $this->get('pluginlogger')->info(
                'StripePayment: Creating charge for order upon receiving "source.chargeable" webhook event',
                [
                    'orderId' => $order->getId(),
                    'eventId' => $event->id,
                ]
            );
            $charge = $this->createChargeForOrder($source, $order);
            $order = $this->updateOrderWithCharge($charge);
```